### ISSUE TITLE
Force invalidation on client

### ISSUE THREAD

**Issue by fabien** [dwstech/muster#27]
**Is your feature request related to a problem? Please describe.**

When using muster-remote over http, it's currently not possible to trigger subscriptions to re-fetch from a remote instance:

```javascript
const clientApp = muster({
    server: remote('http://localhost:8090/api')
});

let q = query(ref('server', 'users'), entries({
    id: true, fullName: true, active: true
}));

clientApp.resolve(q).subscribe(function(result) {
    console.log(JSON.stringify(result, null, 4));
});

// modify the remote data using somehow (set(), ajax call, ...) from the client,
// and force invalidation. The following won't work:

clientApp.resolve(invalidate()).then(function() { // neither does: invalidate(ref('server'))
    console.log('Invalidated');
});
```
Perhaps the invalidation does work, it just doesn't emit this to active subscribers. The websocket-based connection would probably work (haven't tested yet), but a generic way to invalidate remote graphs should work, too. 

Note that this is particularly the case when querying items (arrays of items) with `entries()`. This seems to work: `query(ref('server', 'usersById', 'kate'))` - subscriptions are re-fetched in this case. 

**Describe the solution you'd like**

I would like `invalidate()` to consistently invalidate and emit (and perhaps even forced - as an option to `invalidate()`), so any subscriptions are re-evaluated.

**Comment by tmikus** [comment]
Hi @fabien,

I just tried replicating the described problem in a unit test, but I was unable to reproduce your result. In my case it all worked fine. See the unit test as a reference:
```javascript
runScenario(() => {
    let remoteInstance: Muster;
    return {
      description: 'GIVEN a remote graph with a user data',
      before: () => {
        remoteInstance = muster({
          user: variable(toNode({
            firstName: 'Bob',
            lastName: 'Smith',
          })),
        });
        mockRemoteGraph(remoteInstance);
      },
      graph: () => muster({
        remote: remote('http://test.url'),
      }),
      operations: [
        operation({
          description: 'WHEN making a query to get the user from the server',
          before() {
            jest.clearAllMocks();
          },
          input: query(ref('remote', 'user'), {
            firstName: true,
            lastName: true,
          }),
          expected: value({
            firstName: 'Bob',
            lastName: 'Smith',
          }),
          assert() {
            assertRemoteRequest();
          },
          operations: (subscriber) => [
            operation({
              description: 'AND the server branch changes',
              async before() {
                jest.clearAllMocks();
                await remoteInstance.resolve(set('user', toNode({
                  firstName: 'Kate',
                  lastName: 'Doe',
                })));
              },
              assert() {
                expect(subscriber().next).not.toHaveBeenCalled();
              },
              operations: [
                operation({
                  description: 'AND then the proxy is invalidated',
                  before() {
                    jest.clearAllMocks();
                  },
                  input: invalidate(ref('remote')),
                  assert() {
                    assertRemoteRequest();
                    expect(subscriber().next).toHaveBeenCalledWith(value({
                      firstName: 'Kate',
                      lastName: 'Doe',
                    }));
                  },
                }),
              ],
            }),
          ],
        }),
      ],
    };
  });
```

Could you provide me with a reproducible example of the mentioned bug?

**Comment by fabien** [comment]
Hi @tmikus thanks for checking this. You are correct, subscribing to a specific item (= single object) works (it has been for me, too). However, subscribing to a remote that returns an array asynchronously has not been working for me.

So in your example, you would query and update (change the firstName for example) an item in an array of users, and observe `query(ref(‘remote’, ‘users’), entries({ firstName: true, lastName: true })`. Hope this helps!

**Comment by fabien** [comment]
Unfortunately, I have not been able to get the invalidation on the remote side working. Turns out that the data isn't actually invalidated, unless I'm subscribing to it on the server-side as well. The server caches the result, and it's never updated. The same result is returned for each connection made (so it's not being cached on the client either).

**Comment by fabien** [comment]
I'm starting to think that `invalidate()` does not automatically invalidate child nodes - perhaps this explains why I haven't been able to invalidate a virtual collection (object with dynamic key matching)?

**Comment by dws-tomasz** [comment]
Hi @fabien, sorry for lack of communication - just came back from holiday. I managed to reproduce the bug with invalidating an array. It looks like something we missed in our tests, and which should work. You can expect a fix for this shortly.

**Comment by fabien** [comment]
Hi @dws-tomasz, no worries! Thanks for looking into this, and great to hear that you've been able to reproduce the issue on your side.

**Comment by dws-dave** [comment]
Closed via #51
